server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,183 @@
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TransportGetFromTranslogAction extends HandledTransportAction<
+    TransportGetFromTranslogAction.Request,
+    TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, Request::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final GetRequest getRequest = request.getRequest();
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert getRequest.realtime();
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
+        var get = new Engine.Get(getRequest.realtime(), getRequest.realtime(), getRequest.id()).version(getRequest.version())
+            .versionType(getRequest.versionType());
+        assert Objects.equals(get.uid().field(), IdFieldMapper.NAME) : get.uid().field();
+        ActionListener.completeWith(listener, () -> {
+            if (engine.isInVersionMap(get.uid().bytes())) {
+                var result = indexShard.getService()
+                    .get(
+                        getRequest.id(),
+                        getRequest.storedFields(),
+                        getRequest.realtime(),
+                        getRequest.version(),
+                        getRequest.versionType(),
+                        getRequest.fetchSourceContext(),
+                        getRequest.isForceSyntheticSource()
+                    );
+                return new Response(result, -1);
+            }
+            long segmentGeneration = engine.commitStats().getGeneration();
```

Comment:
I think there is a problem here - and perhaps also in the `RefreshResult`'s generation (will need to dig a bit into that).

If we cannot get rid of all regular refresh'es, a normal refresh will drop the now refreshed state from the version map, but will not advance the generation (which only advances on lucene commit/elasticsearch flush).

I see two ways out:

1. Convert all refresh'es to flush in stateless. This probably requires us to disable scheduled refresh (if not already).
2. Track that we cleared the version map due to a refresh and not due to a flush. Then trigger a flush here to remedy it. We could add a engine.getRefreshGeneration() method that returns commit generation + 1 if the version map has had a refresh.

I'm inclined to go with 2.

Reply:
We discussed this some more and the fact that a flush has an embedded refresh makes option 2 difficult. We will instead track the ids affected since the last flush. We may need to only do this when tracking translog location and when we are on a promotable non-search shard. Also, reversing whether we track translog location could be beneficial in order to keep memory use at bay.

Reply:
For now I'll park this PR. This needs to be continued once we have the mechanism to keep track of the evacuated IDs due to a refresh, in LiveVersionMap. I'll create a follow up Jira issue for that.